Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs

```diff
@@ -102,42 +103,39 @@ public void SetData_FromOtherThread_DoesNotOverride ()
     public void SetData_OnThreadPoolThread_IsResetForNextWorkItem ()
     {
       object result = string.Empty;
-      int threadId1 = -1;
-      int threadId2 = -2;
-      ThreadPool.GetMinThreads (out var minWorkerThreads, out var minCompletionPortThreads);
-      ThreadPool.GetMaxThreads (out var maxWorkerThreads, out var maxCompletionPortThreads);
-      ThreadPool.GetAvailableThreads (out var availableWorkerThreads, out _);
-      try
-      {
-        ThreadPool.SetMinThreads (1, 0);
-        ThreadPool.SetMaxThreads (maxWorkerThreads - availableWorkerThreads + 1, 0);
-        using var resetEvent = new AutoResetEvent (false);
 
+      using var resetEvent = new AutoResetEvent (false);
+
+      var visitedThreadPoolThreads = new HashSet<int>();
+      var completed = false;
+
+      while (!completed)
+      {
         ThreadPool.QueueUserWorkItem (
             _ =>
             {
-              threadId1 = Thread.CurrentThread.ManagedThreadId;
-              _provider.SetData (c_testKey, "1");
-              resetEvent.Set();
-            });
-        resetEvent.WaitOne();
-        ThreadPool.QueueUserWorkItem (
-            _ =>
-            {
-              threadId2 = Thread.CurrentThread.ManagedThreadId;
-              result = _provider.GetData (c_testKey);
+              var threadPoolThreadId = Thread.CurrentThread.ManagedThreadId;
+              var didNotVisitThreadPoolThreadYet = visitedThreadPoolThreads.Add (threadPoolThreadId);
+
+              if (didNotVisitThreadPoolThreadYet)
+              {
+                _provider.SetData (c_testKey, "1");
+              }
+              else
+              {
+                result = _provider.GetData (c_testKey);
+                completed = true;
+              }
+
+              Thread.MemoryBarrier();
               resetEvent.Set();
             });
-        resetEvent.WaitOne();
-      }
-      finally
-      {
-        ThreadPool.SetMaxThreads (maxWorkerThreads, maxCompletionPortThreads);
-        ThreadPool.SetMinThreads (minWorkerThreads, minCompletionPortThreads);
+
+        if (!resetEvent.WaitOne (3 * 1000))
```

Comment:
```suggestion
        if (!resetEvent.WaitOne (TimeSpan.FromSeconds (3)))
```

Yes, I know, highly unlikely. But a hanging build on the build server trumps all the 'unlikelies' we can see and easily prevent. The problem is, first you need to notice, then you need to be able to attach a debugger before someone kills it, and then you have a chance to figure it out. Unfortunately, the in-general scenario of the hanging build without obivous cause isn't just theory, so going the extra inch is usually worth the effort. Especially if the lead reviewer is usually the one to find the hanging build early in the morning and then gets to have some fun.